File: packages/altimate-engine/src/altimate_engine/credential_store.py
from __future__ import annotations

from typing import Any

SERVICE_NAME = "altimate-code"

SENSITIVE_FIELDS = {
    "password",
    "private_key_passphrase",
    "access_token",
    "ssh_password",
    "connection_string",
}

_keyring_cache: bool | None = None


def _keyring_available() -> bool:
    global _keyring_cache
    if _keyring_cache is not None:
        return _keyring_cache
    try:
        import keyring  # noqa: F401
        _keyring_cache = True
    except ImportError:
        _keyring_cache = False
    return _keyring_cache


def store_credential(name: str, field: str, value: str) -> bool:
    if not _keyring_available():
        return False
    import keyring
    try:
        keyring.set_password(SERVICE_NAME, f"{name}/{field}", value)
        return True
    except Exception:
        return False


def get_credential(name: str, field: str) -> str | None:
    if not _keyring_available():
        return None
    import keyring
    try:
        return keyring.get_password(SERVICE_NAME, f"{name}/{field}")
    except Exception:
        return None


def delete_all_credentials(name: str) -> None:
    if not _keyring_available():
        return
    import keyring
    for field in SENSITIVE_FIELDS:
        try:
            keyring.delete_password(SERVICE_NAME, f"{name}/{field}")
        except Exception:
            pass


def resolve_config(name: str, config: dict[str, Any]) -> dict[str, Any]:
    resolved = dict(config)
    for field in SENSITIVE_FIELDS:
        if resolved.get(field) is None:
            cred = get_credential(name, field)
            if cred is not None:
                resolved[field] = cred
    return resolved


def save_connection(name: str, config: dict[str, Any], config_path: str | None = None) -> dict[str, Any]:
    from pathlib import Path
    import json

    if config_path is None:
        config_path = str(Path.home() / ".altimate-code" / "connections.json")

    path = Path(config_path)
    path.parent.mkdir(parents=True, exist_ok=True)

    existing = {}
    if path.exists():
        with open(path) as f:
            existing = json.load(f)

    safe_config = {k: v for k, v in config.items() if k not in SENSITIVE_FIELDS}
    for field in SENSITIVE_FIELDS:
        if field in config and config[field] is not None:
            store_credential(name, field, str(config[field]))
            safe_config[field] = None

    existing[name] = safe_config
    with open(path, "w") as f:
        json.dump(existing, f, indent=2)

    return safe_config


def remove_connection(name: str, config_path: str | None = None) -> bool:
    from pathlib import Path
    import json

    if config_path is None:
        config_path = str(Path.home() / ".altimate-code" / "connections.json")

    path = Path(config_path)
    if not path.exists():
        return False

    with open(path) as f:
        existing = json.load(f)

    if name not in existing:
        return False

    del existing[name]
    delete_all_credentials(name)

    with open(path, "w") as f:
        json.dump(existing, f, indent=2)

    return True

File: packages/opencode/src/altimate/native/connections/credential-store.ts
/**
 * Credential management for connection configs.
 *
 * 3-tier fallback:
 * 1. keytar (OS Keychain) — preferred, secure
 * 2. ALTIMATE_CODE_CONN_* env vars — for headless/CI environments
 * 3. Refuse — never store plaintext credentials in config JSON
 */

import { Log } from "../../../util/log"
import type { ConnectionConfig } from "@altimateai/drivers"

const SERVICE_NAME = "altimate-code"

const SENSITIVE_FIELDS = new Set([
  "password",
  "private_key",
  "privateKey",
  "private_key_passphrase",
  "privateKeyPassphrase",
  "privateKeyPass",
  "access_token",
  "token",
  "oauth_client_secret",
  "oauthClientSecret",
  "passcode",
  "ssh_password",
  "connection_string",
  "credentials_json",
  "keyfile_json",
  "ssl_key",
  "ssl_cert",
  "ssl_ca",
  "tls_key",
  "tls_cert",
  "tls_ca_cert",
])

/** Cached keytar module (or null if unavailable). */
let keytarModule: any | null | undefined = undefined

async function getKeytar(): Promise<any | null> {
  if (keytarModule !== undefined) return keytarModule
  try {
    // @ts-expect-error — optional dependency, loaded at runtime
    keytarModule = await import("keytar")
    return keytarModule
  } catch {
    Log.Default.warn(
      "keytar not available — use ALTIMATE_CODE_CONN_* env vars for secure credential storage",
    )
    keytarModule = null
    return null
  }
}

/** Store a single credential in the OS keychain (or return false if unavailable). */
export async function storeCredential(
  connectionName: string,
  field: string,
  value: string,
): Promise<boolean> {
  const keytar = await getKeytar()
  if (!keytar) return false
  const account = `${connectionName}/${field}`
  await keytar.setPassword(SERVICE_NAME, account, value)
  return true
}

/** Retrieve a single credential from the OS keychain (or return null). */
export async function getCredential(
  connectionName: string,
  field: string,
): Promise<string | null> {
  const keytar = await getKeytar()
  if (!keytar) return null
  const account = `${connectionName}/${field}`
  return keytar.getPassword(SERVICE_NAME, account)
}

/** Delete a single credential from the OS keychain. */
export async function deleteCredential(
  connectionName: string,
  field: string,
): Promise<boolean> {
  const keytar = await getKeytar()
  if (!keytar) return false
  const account = `${connectionName}/${field}`
  return keytar.deletePassword(SERVICE_NAME, account)
}

/**
 * Resolve a connection config by pulling sensitive fields from the keychain.
 * If keytar is unavailable, returns the config as-is (credentials stay in JSON).
 */
export async function resolveConfig(
  name: string,
  config: ConnectionConfig,
): Promise<ConnectionConfig> {
  const resolved = { ...config }
  for (const field of SENSITIVE_FIELDS) {
    if (resolved[field]) continue // already present in config
    const stored = await getCredential(name, field)
    if (stored) {
      resolved[field] = stored
    }
  }
  return resolved
}

/**
 * Save a connection config, extracting sensitive fields to the keychain.
 * Returns the sanitized config and any warnings about stripped credentials.
 */
export async function saveConnection(
  name: string,
  config: ConnectionConfig,
): Promise<{ sanitized: ConnectionConfig; warnings: string[] }> {
  const sanitized = { ...config }
  const warnings: string[] = []
  for (const field of SENSITIVE_FIELDS) {
    const value = config[field]
    if (typeof value !== "string" || !value) continue
    const stored = await storeCredential(name, field, value)
    if (stored) {
      delete sanitized[field]
    } else {
      // keytar unavailable — strip sensitive field from config to prevent
      // plaintext storage. Users should use ALTIMATE_CODE_CONN_* env vars.
      const warning = `Cannot securely store '${field}' for connection '${name}'. ` +
        `Set ALTIMATE_CODE_CONN_${name.toUpperCase()} env var with full config JSON instead.`
      Log.Default.warn(warning)
      warnings.push(warning)
      delete sanitized[field]
    }
  }
  return { sanitized, warnings }
}

/** Check if a field is sensitive. */
export function isSensitiveField(field: string): boolean {
  return SENSITIVE_FIELDS.has(field)
}
